src/dragAndSnapManager.ts

```diff
@@ -0,0 +1,170 @@
+    this._eventManager = eventManager;
+    this._logger = logger;
+    this._onPositionChanged = fn;
+  }
+
+  // Drag and snap public API
+  public setDraggableContainer = (draggableContainer: HTMLDivElement) => {
+    this._draggableContainer = draggableContainer;
+    this._addListeners();
+  };
+
+  public setGuiClientRect = (gueClientRec: GuiClientRect) => {
+    this._gueClientRec = gueClientRec;
+  };
+
+  public onStartDrag = (fn: () => void) => {
+    this._onStartDrag = fn;
+  }
+
+  public onStopDrag = (fn: () => void) => {
+    this._onStopDrag = fn;
+  }
+
+  private _addListeners = () => {
+    this._eventManager.listen(this._draggableContainer!, DragEvents.MouseDown, e => {
+      this._startDrag(e, DragEvents.MouseMove, DragEvents.MouseUp);
+    });
+    this._eventManager.listen(this._draggableContainer!, DragEvents.TouchStart, e => {
+      this._eventManager.unlisten(this._draggableContainer!, DragEvents.MouseDown);
+      this._startDrag(e, DragEvents.TouchMove, DragEvents.TouchEnd);
+    });
+  };
+
+  private _startDrag(e: MouseEvent | TouchEvent, moveEventName: string, endEventName: string) {
+    this._eventManager.listenOnce(document, endEventName, () => {
+      this._stopDrag();
+    });
+
+    // get the mouse cursor position at startup:
+    this._currMousePos.x = getClientX(e);
+    this._currMousePos.y = getClientY(e);
+
+    this._eventManager.listen(document, moveEventName, this._moveDrag);
+    this._eventManager.listenOnce(document, moveEventName, this._onStartDrag);
+  }
+
+  private _moveDrag = (e: MouseEvent | TouchEvent) => {
+    if (this._throttleWait) {
+      return;
+    };
+    e = e || window.event;
+    // calculate the new cursor position:
+    const deltaMousePosX = this._currMousePos.x - getClientX(e);
+    const deltaMousePosY = this._currMousePos.y - getClientY(e);
+    this._currMousePos.x = getClientX(e);
+    this._currMousePos.y = getClientY(e);
+    // set the element's new position
+    if (this._draggableContainer && this._gueClientRec) {
+      let top = parseInt(this._draggableContainer.style.top, 10);
+      let bottom = parseInt(this._draggableContainer.style.bottom, 10);
+      let right = parseInt(this._draggableContainer.style.right, 10);
+      let left = parseInt(this._draggableContainer.style.left, 10);
+
+      if (Number.isInteger(right)) {
+        right = right + deltaMousePosX;
+      } else {
+        left = left - deltaMousePosX;
+      }
+      if (Number.isInteger(bottom)) {
+        bottom = bottom + deltaMousePosY;
+      } else {
+        top = top - deltaMousePosY;
+      }
+
+      this._draggableContainer.style.top = makeStyleString(top);
+      this._draggableContainer.style.right = makeStyleString(right);
+      this._draggableContainer.style.bottom = makeStyleString(bottom);
+      this._draggableContainer.style.left = makeStyleString(left);
```

Comment:
You always apply all four of them? Not better to push it to the "if" before so only the relevant styles will be applied?

Reply:
nice!
implemented